File: Sources/APTradeApp/NewsViewModel.swift
import Foundation
import APTradeApplication
import APTradeDomain

@MainActor
@Observable
final class NewsViewModel {
    private(set) var articles: [NewsArticle] = []
    private(set) var bookmarkedIDs: Set<String> = []
    private(set) var isLoading = false
    let keyMissing: Bool
    var category: NewsCategory = .general
    var filter: String = ""
    var showingSaved = false

    private let fetchMarketNews: FetchMarketNewsUseCase
    private let loadBookmarks: LoadBookmarksUseCase
    private let toggleBookmarkUseCase: ToggleBookmarkUseCase
    private var bookmarks: [NewsArticle] = []

    init(fetchMarketNews: FetchMarketNewsUseCase,
         loadBookmarks: LoadBookmarksUseCase,
         toggleBookmark: ToggleBookmarkUseCase,
         keyMissing: Bool) {
        self.fetchMarketNews = fetchMarketNews
        self.loadBookmarks = loadBookmarks
        self.toggleBookmarkUseCase = toggleBookmark
        self.keyMissing = keyMissing
    }

    /// The list to render: the saved set or the live feed, narrowed by the filter text.
    var visibleArticles: [NewsArticle] {
        let base = showingSaved ? bookmarks : articles
        let query = filter.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard !query.isEmpty else { return base }
        return base.filter {
            $0.headline.lowercased().contains(query) || $0.source.lowercased().contains(query)
        }
    }

    func onAppear() async {
        refreshBookmarks()
        if articles.isEmpty && !keyMissing { await load() }
    }

    func load() async {
        guard !keyMissing else { return }
        isLoading = true
        articles = await fetchMarketNews(category: category)
        isLoading = false
    }

    func setCategory(_ newValue: NewsCategory) async {
        guard newValue != category else { return }
        category = newValue
        await load()
    }

    func toggleBookmark(_ article: NewsArticle) {
        bookmarks = toggleBookmarkUseCase(article)
        bookmarkedIDs = Set(bookmarks.map(\.id))
    }

    func isBookmarked(_ article: NewsArticle) -> Bool { bookmarkedIDs.contains(article.id) }

    private func refreshBookmarks() {
        bookmarks = loadBookmarks()
        bookmarkedIDs = Set(bookmarks.map(\.id))
    }
}

File: desktopApp/src/main/kotlin/com/aptrade/desktop/news/NewsViewModel.kt
package com.aptrade.desktop.news

import com.aptrade.shared.application.FetchMarketNews
import com.aptrade.shared.application.LoadBookmarks
import com.aptrade.shared.application.ToggleBookmark
import com.aptrade.shared.domain.NewsArticle
import com.aptrade.shared.domain.NewsCategory
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class NewsUiState(
    val articles: List<NewsArticle> = emptyList(),
    val bookmarks: List<NewsArticle> = emptyList(),
    val category: NewsCategory = NewsCategory.General,
    val showingSaved: Boolean = false,
    val filter: String = "",
    val isLoading: Boolean = false,
    val keyMissing: Boolean = false,
) {
    /** Ids of bookmarked articles — derived so the UI can render a filled/hollow bookmark
     *  glyph per row without scanning the list. */
    val bookmarkedIds: Set<String> get() = bookmarks.mapTo(HashSet()) { it.id }

    /** macOS parity: base is the saved list when [showingSaved] else the fetched articles;
     *  a blank filter yields the base unchanged, otherwise a live (no-debounce),
     *  case-insensitive substring match on headline OR source. */
    val visibleArticles: List<NewsArticle>
        get() {
            val base = if (showingSaved) bookmarks else articles
            val query = filter.trim()
            if (query.isEmpty()) return base
            val needle = query.lowercase()
            return base.filter {
                it.headline.lowercase().contains(needle) || it.source.lowercase().contains(needle)
            }
        }
}

/** Owns the News tab: category-scoped market news, the Saved toggle, live text filter, and
 *  bookmark toggling. [fetchMarketNews] is null when no news key is configured (AppGraph
 *  passes null) — that becomes [NewsUiState.keyMissing] and no fetch is ever attempted.
 *  Bookmarks always load regardless of key state.
 *  `scope` MUST be single-thread-confined (Dispatchers.Main on desktop): state updates
 *  rely on that confinement instead of locks. */
class NewsViewModel(
    private val fetchMarketNews: FetchMarketNews?,
    private val loadBookmarks: LoadBookmarks,
    private val toggleBookmark: ToggleBookmark,
    private val scope: CoroutineScope,
) {
    private val _state = MutableStateFlow(NewsUiState(keyMissing = fetchMarketNews == null))
    val state: StateFlow<NewsUiState> = _state

    fun start() {
        scope.launch {
            _state.update { it.copy(bookmarks = loadBookmarks.execute()) }
        }
        val fetch = fetchMarketNews ?: return
        scope.launch { load(fetch, _state.value.category) }
    }

    fun setCategory(category: NewsCategory) {
        if (_state.value.category == category) return
        _state.update { it.copy(category = category) }
        val fetch = fetchMarketNews ?: return
        scope.launch { load(fetch, category) }
    }

    fun setShowingSaved(showingSaved: Boolean) =
        _state.update { it.copy(showingSaved = showingSaved) }

    fun setFilter(filter: String) = _state.update { it.copy(filter = filter) }

    fun refresh() {
        val fetch = fetchMarketNews ?: return
        scope.launch { load(fetch, _state.value.category) }
    }

    fun toggleBookmark(article: NewsArticle) {
        scope.launch {
            try {
                val updated = toggleBookmark.execute(article)
                _state.update { it.copy(bookmarks = updated) }
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                // bookmark persistence is best-effort; a failure leaves the last-good list
            }
        }
    }

    private suspend fun load(fetch: FetchMarketNews, category: NewsCategory) {
        _state.update { it.copy(isLoading = true) }
        // FetchMarketNews already swallows failures to an empty list; the only throw that
        // escapes it is CancellationException, which must propagate to cancel this load.
        try {
            val articles = fetch.execute(category)
            _state.update { it.copy(articles = articles, isLoading = false) }
        } catch (e: CancellationException) {
            throw e
        }
    }
}
